File: src/shipit/providers/staticfile.py
import json
import re
import shlex
from dataclasses import dataclass
from functools import cached_property
from pathlib import Path
from typing import Dict, Optional

from tomlkit import aot, document, table
import yaml
from pydantic_settings import SettingsConfigDict

from .base import (
    DetectResult,
    DependencySpec,
    Provider,
    _exists,
    MountSpec,
    ServiceSpec,
    VolumeSpec,
    Config,
)


class StaticFileConfig(Config):
    model_config = SettingsConfigDict(extra="ignore", env_prefix="SHIPIT_")

    convert_redirects: bool = True
    sws_version: Optional[str] = "2.38.0"
    static_dir: Optional[str] = None
    # Rendered sws.toml redirects (from a _redirects file), computed at load
    # time so the Starlark provider stays filesystem-free.
    redirects_config: Optional[str] = None


@dataclass(frozen=True)
class RedirectRule:
    source: str
    destination: str
    kind: int


class StaticFileProvider:
    REDIRECTS_CONFIG_FILE = "sws.toml"
    REDIRECTS_CONFIG_MOUNT = "static_config"
    REDIRECTS_SOURCE = "_redirects"
    REDIRECT_STATUS_CODES = {301, 302}
    _PARAM_PATTERN = re.compile(r":([A-Za-z][A-Za-z0-9_]*)")
    _SOURCE_TOKEN_PATTERN = re.compile(r":([A-Za-z][A-Za-z0-9_]*)|\*")

    config: Optional[dict] = None
    path: Path

    def __init__(self, path: Path, config: StaticFileConfig):
        self.path = path
        self.config = config

    @classmethod
    def load_config(
        cls, path: Path, base_config: Config
    ) -> StaticFileConfig:
        config = cls._load_static_config(path, base_config)
        config.redirects_config = compute_redirects_config(
            path, config.static_dir, config.convert_redirects
        )
        return config

    @classmethod
    def _load_static_config(
        cls, path: Path, base_config: Config
    ) -> StaticFileConfig:
        if (path / "Staticfile").exists():
            config = None
            try:
                config = yaml.safe_load((path / "Staticfile").read_text())
            except yaml.YAMLError as e:
                print(f"Error loading Staticfile: {e}")
                pass

            if config:
                return StaticFileConfig(
                    **base_config.model_dump(),
                    static_dir=config.get("root"),
                )
        if _exists(path, "public/index.html") or _exists(path, "public/index.htm"):
            return StaticFileConfig(static_dir="public", **base_config.model_dump())

        return StaticFileConfig(**base_config.model_dump())

    @classmethod
    def name(cls) -> str:
        return "staticfile"

    @classmethod
    def detect(
        cls, path: Path, config: Config
    ) -> Optional[DetectResult]:
        is_python_php_js_project = _exists(
            path, "package.json", "pyproject.toml", "composer.json"
        )
        if _exists(path, "Staticfile"):
            return DetectResult(cls.name(), 50)
        if not is_python_php_js_project:
            if _exists(
                path, "index.html", "index.htm", "public/index.htm", "public/index.html"
            ):
                return DetectResult(cls.name(), 10)
            return DetectResult(cls.name(), 10)
        if config.commands.start and config.commands.start.startswith(
            "static-web-server "
        ):
            return DetectResult(cls.name(), 70)
        return None

    def dependencies(self) -> list[DependencySpec]:
        return [
            DependencySpec(
                "static-web-server",
                var_name="config.sws_version",
                use_in_serve=True,
            )
        ]

    def build_steps_redirects(self) -> list[str]:
        redirects_config = self.redirects_config
        if not redirects_config:
            return []
        return [
            'write("{}/%s".format(static_config.path), %s)'
            % (
                self.REDIRECTS_CONFIG_FILE,
                json.dumps(redirects_config),
            )
        ]

    def build_steps(self) -> list[str]:
        source = json.dumps(self.config.static_dir or ".")
        if self.config.app_subdir:
            if self.config.static_dir:
                source = f'"{{}}/{self.config.static_dir}".format(app_subdir)'
            else:
                source = "app_subdir"
        return [
            'workdir(static_app.path)',
            f'copy({source}, ".", ignore=[".git"])',
        ] + self.build_steps_redirects()

    def prepare_steps(self) -> Optional[list[str]]:
        return None

    def declarations(self) -> Optional[str]:
        return None

    def commands(self) -> Dict[str, str]:
        if self.redirects_config:
            return {
                "start": '"static-web-server --root={} --log-level=info --config-file={}/%s --port={}".format(static_app.serve_path, static_config.serve_path, PORT)'
                % self.REDIRECTS_CONFIG_FILE
            }
        return {
            "start": '"static-web-server --root={} --log-level=info --port={}".format(static_app.serve_path, PORT)'
        }

    def mounts(self) -> list[MountSpec]:
        mounts = [MountSpec("static_app")]
        if self.redirects_config:
            mounts.append(MountSpec(self.REDIRECTS_CONFIG_MOUNT))
        return mounts

    def volumes(self) -> list[VolumeSpec]:
        return []

    def env(self) -> Optional[Dict[str, str]]:
        return None

    def services(self) -> list[ServiceSpec]:
        return []

    @cached_property
    def redirects_config(self) -> Optional[str]:
        return compute_redirects_config(
            self.path, self.config.static_dir, self.config.convert_redirects
        )

    @classmethod
    def _load_redirect_rules(cls, redirects_path: Path) -> list[RedirectRule]:
        rules: list[RedirectRule] = []
        for line_number, raw_line in enumerate(
            redirects_path.read_text().splitlines(), start=1
        ):
            line = raw_line.strip()
            if not line or line.startswith("#"):
                continue

            try:
                parts = shlex.split(line)
            except ValueError as exc:
                raise ValueError(
                    f"{redirects_path}:{line_number}: invalid _redirects rule"
                ) from exc

            if len(parts) < 2:
                raise ValueError(
                    f"{redirects_path}:{line_number}: expected source and "
                    "destination"
                )

            source, destination, *rest = parts
            kind = 301
            if rest and rest[0].isdigit():
                kind = int(rest[0])
                rest = rest[1:]

            if kind not in cls.REDIRECT_STATUS_CODES:
                raise ValueError(
                    f"{redirects_path}:{line_number}: redirect status {kind} "
                    "is not supported by static-web-server"
                )

            if rest:
                raise ValueError(
                    f"{redirects_path}:{line_number}: conditions and forced "
                    "redirects are not supported"
                )

            sws_source, replacements = cls._translate_source(
                redirects_path, line_number, source
            )
            sws_destination = cls._translate_destination(
                redirects_path, line_number, destination, replacements
            )
            rules.append(
                RedirectRule(
                    source=sws_source,
                    destination=sws_destination,
                    kind=kind,
                )
            )

        return rules

    @classmethod
    def _translate_source(
        cls, redirects_path: Path, line_number: int, source: str
    ) -> tuple[str, dict[str, int]]:
        if "://" in source:
            raise ValueError(
                f"{redirects_path}:{line_number}: redirect sources must be "
                "local paths"
            )
        if "?" in source:
            raise ValueError(
                f"{redirects_path}:{line_number}: query matching is not "
                "supported"
            )
        if not source.startswith("/"):
            raise ValueError(
                f"{redirects_path}:{line_number}: redirect sources must start "
                "with '/'"
            )

        translated_parts: list[str] = []
        replacements: dict[str, int] = {}
        last_index = 0
        next_index = 1

        for match in cls._SOURCE_TOKEN_PATTERN.finditer(source):
            translated_parts.append(source[last_index : match.start()])
            param_name = match.group(1)
            if param_name is not None:
                if param_name in replacements:
                    raise ValueError(
                        f"{redirects_path}:{line_number}: duplicate source "
                        f"parameter :{param_name}"
                    )
                replacements[param_name] = next_index
                translated_parts.append("{*}")
            else:
                if "splat" in replacements:
                    raise ValueError(
                        f"{redirects_path}:{line_number}: only one splat "
                        "segment is supported"
                    )
                replacements["splat"] = next_index
                translated_parts.append("{**}")
            next_index += 1
            last_index = match.end()

        translated_parts.append(source[last_index:])
        return "".join(translated_parts), replacements

    @classmethod
    def _translate_destination(
        cls,
        redirects_path: Path,
        line_number: int,
        destination: str,
        replacements: dict[str, int],
    ) -> str:
        if "*" in destination:
            raise ValueError(
                f"{redirects_path}:{line_number}: destination splats must use "
                ":splat"
            )

        def replace_param(match: re.Match[str]) -> str:
            param_name = match.group(1)
            if param_name not in replacements:
                raise ValueError(
                    f"{redirects_path}:{line_number}: destination references "
                    f"unknown parameter :{param_name}"
                )
            return f"${replacements[param_name]}"

        return cls._PARAM_PATTERN.sub(replace_param, destination)


def compute_redirects_config(
    path: Path, static_dir: Optional[str], convert_redirects: bool = True
) -> Optional[str]:
    """Render a _redirects file into sws.toml redirect rules (or None)."""
    if not convert_redirects:
        return None

    redirects_path = path / StaticFileProvider.REDIRECTS_SOURCE
    if static_dir:
        static_dir_redirects = path / static_dir / StaticFileProvider.REDIRECTS_SOURCE
        if static_dir_redirects.is_file():
            redirects_path = static_dir_redirects
    if not redirects_path.is_file():
        return None

    rules = StaticFileProvider._load_redirect_rules(redirects_path)
    if not rules:
        return None

    doc = document()
    advanced = table()
    redirects = aot()
    for rule in rules:
        entry = table()
        entry.add("source", rule.source)
        entry.add("destination", rule.destination)
        entry.add("kind", rule.kind)
        redirects.append(entry)

    advanced.add("redirects", redirects)
    doc.add("advanced", advanced)
    return doc.as_string()

File: crates/shipit-providers/src/staticfile.rs
//! Port of `src/shipit/providers/staticfile.py`.
//!
//! Static site detection, Staticfile/`public/` docroot resolution, and the
//! `_redirects` → sws.toml redirects pipeline. The rendered TOML string
//! must match Python's tomlkit output byte-for-byte (it ends up in plans).

use std::collections::HashMap;
use std::path::Path;
use std::sync::OnceLock;

use anyhow::{anyhow, bail, Context, Result};
use regex::Regex;
use serde::{Deserialize, Serialize};

use crate::base::{env_bool, env_str, BaseConfig, DetectResult, HasBase};

pub const NAME: &str = "staticfile";

pub const REDIRECTS_SOURCE: &str = "_redirects";
pub const REDIRECT_STATUS_CODES: [i64; 2] = [301, 302];

fn param_pattern() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r":([A-Za-z][A-Za-z0-9_]*)").expect("valid regex"))
}

fn source_token_pattern() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r":([A-Za-z][A-Za-z0-9_]*)|\*").expect("valid regex"))
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StaticFileConfig {
    #[serde(flatten)]
    pub base: BaseConfig,
    pub convert_redirects: bool,
    pub sws_version: Option<String>,
    pub static_dir: Option<String>,
    /// Rendered sws.toml redirects (from a _redirects file), computed at
    /// load time so the Starlark provider stays filesystem-free.
    pub redirects_config: Option<String>,
}

impl Default for StaticFileConfig {
    fn default() -> Self {
        Self {
            base: BaseConfig::default(),
            convert_redirects: true,
            sws_version: Some("2.38.0".to_owned()),
            static_dir: None,
            redirects_config: None,
        }
    }
}

impl HasBase for StaticFileConfig {
    fn base(&self) -> &BaseConfig {
        &self.base
    }
    fn base_mut(&mut self) -> &mut BaseConfig {
        &mut self.base
    }
}

impl StaticFileConfig {
    /// pydantic-settings construction with only the base fields passed as
    /// init kwargs: the env reaches every staticfile-specific field.
    pub(crate) fn from_env(base: BaseConfig) -> Result<Self> {
        Ok(Self {
            base,
            convert_redirects: env_bool("convert_redirects")?.unwrap_or(true),
            sws_version: env_str("sws_version").or_else(|| Some("2.38.0".to_owned())),
            static_dir: env_str("static_dir"),
            redirects_config: env_str("redirects_config"),
        })
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct RedirectRule {
    pub source: String,
    pub destination: String,
    pub kind: i64,
}

fn exists(path: &Path, candidates: &[&str]) -> bool {
    candidates.iter().any(|c| path.join(c).exists())
}

/// Minimal `yaml.safe_load` replacement for the flat `key: value` YAML
/// documents Shipit reads (Staticfile roots, Jekyll destinations). Only
/// top-level scalar mappings are captured.
pub(crate) fn parse_simple_yaml(contents: &str) -> HashMap<String, String> {
    let mut map = HashMap::new();
    for line in contents.lines() {
        if line.starts_with([' ', '\t']) {
            continue; // nested value; the flat lookups never need them
        }
        let trimmed = line.trim();
        if trimmed.is_empty() || trimmed.starts_with('#') || trimmed.starts_with('-') {
            continue;
        }
        let Some((key, value)) = trimmed.split_once(':') else {
            continue;
        };
        let key = key.trim();
        let mut value = value.trim();
        // Drop trailing comments (approximation of YAML's rules).
        if let Some(index) = value.find(" #") {
            value = value[..index].trim_end();
        }
        let value = value
            .strip_prefix('"')
            .and_then(|v| v.strip_suffix('"'))
            .or_else(|| value.strip_prefix('\'').and_then(|v| v.strip_suffix('\'')))
            .unwrap_or(value);
        if key.is_empty() || value.is_empty() {
            continue;
        }
        map.insert(key.to_owned(), value.to_owned());
    }
    map
}

/// Port of `StaticFileProvider.load_config`.
pub fn load_config(path: &Path, base: BaseConfig) -> Result<StaticFileConfig> {
    let mut config = load_static_config(path, base)?;
    config.redirects_config =
        compute_redirects_config(path, config.static_dir.as_deref(), config.convert_redirects)?;
    Ok(config)
}

/// Port of `StaticFileProvider._load_static_config`.
fn load_static_config(path: &Path, base: BaseConfig) -> Result<StaticFileConfig> {
    let staticfile_path = path.join("Staticfile");
    if staticfile_path.exists() {
        if let Ok(contents) = std::fs::read_to_string(&staticfile_path) {
            let parsed = parse_simple_yaml(&contents);
            if !parsed.is_empty() {
                // static_dir is an explicit init kwarg here (even when the
                // Staticfile has no root), so the env never overrides it.
                let mut config = StaticFileConfig::from_env(base)?;
                config.static_dir = parsed.get("root").cloned();
                return Ok(config);
            }
        }
    }
    if exists(path, &["public/index.html"]) || exists(path, &["public/index.htm"]) {
        let mut config = StaticFileConfig::from_env(base)?;
        config.static_dir = Some("public".to_owned());
        return Ok(config);
    }
    StaticFileConfig::from_env(base)
}

/// Port of `StaticFileProvider.detect`.
pub fn detect(path: &Path, base: &BaseConfig) -> Option<DetectResult> {
    let is_python_php_js_project =
        exists(path, &["package.json", "pyproject.toml", "composer.json"]);
    if exists(path, &["Staticfile"]) {
        return Some(DetectResult {
            name: NAME,
            score: 50,
        });
    }
    if !is_python_php_js_project {
        // Python returns 10 whether or not an index file exists.
        return Some(DetectResult {
            name: NAME,
            score: 10,
        });
    }
    if base
        .commands
        .start
        .as_deref()
        .is_some_and(|start| start.starts_with("static-web-server "))
    {
        return Some(DetectResult {
            name: NAME,
            score: 70,
        });
    }
    None
}

/// Port of `compute_redirects_config`: render a `_redirects` file into
/// sws.toml redirect rules (or None).
pub fn compute_redirects_config(
    path: &Path,
    static_dir: Option<&str>,
    convert_redirects: bool,
) -> Result<Option<String>> {
    if !convert_redirects {
        return Ok(None);
    }

    let mut redirects_path = path.join(REDIRECTS_SOURCE);
    if let Some(static_dir) = static_dir {
        let static_dir_redirects = path.join(static_dir).join(REDIRECTS_SOURCE);
        if static_dir_redirects.is_file() {
            redirects_path = static_dir_redirects;
        }
    }
    if !redirects_path.is_file() {
        return Ok(None);
    }

    let rules = load_redirect_rules(&redirects_path)?;
    if rules.is_empty() {
        return Ok(None);
    }

    // tomlkit renders the [advanced] super-table implicitly: only the
    // [[advanced.redirects]] blocks appear, separated by blank lines.
    let mut rendered = String::new();
    for (index, rule) in rules.iter().enumerate() {
        if index > 0 {
            rendered.push('\n');
        }
        rendered.push_str("[[advanced.redirects]]\n");
        rendered.push_str(&format!("source = {}\n", toml_basic_string(&rule.source)));
        rendered.push_str(&format!(
            "destination = {}\n",
            toml_basic_string(&rule.destination)
        ));
        rendered.push_str(&format!("kind = {}\n", rule.kind));
    }
    Ok(Some(rendered))
}

/// tomlkit-style basic string rendering.
fn toml_basic_string(value: &str) -> String {
    let mut out = String::with_capacity(value.len() + 2);
    out.push('"');
    for ch in value.chars() {
        match ch {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\n' => out.push_str("\\n"),
            '\r' => out.push_str("\\r"),
            '\t' => out.push_str("\\t"),
            '\u{8}' => out.push_str("\\b"),
            '\u{c}' => out.push_str("\\f"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04X}", c as u32)),
            c => out.push(c),
        }
    }
    out.push('"');
    out
}

/// Port of `StaticFileProvider._load_redirect_rules`.
pub(crate) fn load_redirect_rules(redirects_path: &Path) -> Result<Vec<RedirectRule>> {
    let contents = std::fs::read_to_string(redirects_path)
        .with_context(|| format!("{}: unreadable _redirects file", redirects_path.display()))?;
    let mut rules = Vec::new();
    for (index, raw_line) in contents.lines().enumerate() {
        let line_number = index + 1;
        let line = raw_line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        let parts = shlex_split(line).ok_or_else(|| {
            anyhow!(
                "{}:{line_number}: invalid _redirects rule",
                redirects_path.display()
            )
        })?;
        if parts.len() < 2 {
            bail!(
                "{}:{line_number}: expected source and destination",
                redirects_path.display()
            );
        }

        let source = &parts[0];
        let destination = &parts[1];
        let mut rest = &parts[2..];
        let mut kind: i64 = 301;
        if let Some(first) = rest.first() {
            if !first.is_empty() && first.chars().all(|c| c.is_ascii_digit()) {
                kind = first.parse().map_err(|_| {
                    anyhow!(
                        "{}:{line_number}: redirect status {first} is not supported by \
                         static-web-server",
                        redirects_path.display()
                    )
                })?;
                rest = &rest[1..];
            }
        }

        if !REDIRECT_STATUS_CODES.contains(&kind) {
            bail!(
                "{}:{line_number}: redirect status {kind} is not supported by \
                 static-web-server",
                redirects_path.display()
            );
        }
        if !rest.is_empty() {
            bail!(
                "{}:{line_number}: conditions and forced redirects are not supported",
                redirects_path.display()
            );
        }

        let (sws_source, replacements) = translate_source(redirects_path, line_number, source)?;
        let sws_destination =
            translate_destination(redirects_path, line_number, destination, &replacements)?;
        rules.push(RedirectRule {
            source: sws_source,
            destination: sws_destination,
            kind,
        });
    }
    Ok(rules)
}

/// Port of `StaticFileProvider._translate_source`.
fn translate_source(
    redirects_path: &Path,
    line_number: usize,
    source: &str,
) -> Result<(String, HashMap<String, usize>)> {
    if source.contains("://") {
        bail!(
            "{}:{line_number}: redirect sources must be local paths",
            redirects_path.display()
        );
    }
    if source.contains('?') {
        bail!(
            "{}:{line_number}: query matching is not supported",
            redirects_path.display()
        );
    }
    if !source.starts_with('/') {
        bail!(
            "{}:{line_number}: redirect sources must start with '/'",
            redirects_path.display()
        );
    }

    let mut translated = String::new();
    let mut replacements: HashMap<String, usize> = HashMap::new();
    let mut last_index = 0;
    for (next_index, captures) in (1..).zip(source_token_pattern().captures_iter(source)) {
        let whole = captures.get(0).expect("match 0 exists");
        translated.push_str(&source[last_index..whole.start()]);
        if let Some(param) = captures.get(1) {
            let param_name = param.as_str();
            if replacements.contains_key(param_name) {
                bail!(
                    "{}:{line_number}: duplicate source parameter :{param_name}",
                    redirects_path.display()
                );
            }
            replacements.insert(param_name.to_owned(), next_index);
            translated.push_str("{*}");
        } else {
            if replacements.contains_key("splat") {
                bail!(
                    "{}:{line_number}: only one splat segment is supported",
                    redirects_path.display()
                );
            }
            replacements.insert("splat".to_owned(), next_index);
            translated.push_str("{**}");
        }
        last_index = whole.end();
    }

    translated.push_str(&source[last_index..]);
    Ok((translated, replacements))
}

/// Port of `StaticFileProvider._translate_destination`.
fn translate_destination(
    redirects_path: &Path,
    line_number: usize,
    destination: &str,
    replacements: &HashMap<String, usize>,
) -> Result<String> {
    if destination.contains('*') {
        bail!(
            "{}:{line_number}: destination splats must use :splat",
            redirects_path.display()
        );
    }

    let mut translated = String::new();
    let mut last_index = 0;
    for captures in param_pattern().captures_iter(destination) {
        let whole = captures.get(0).expect("match 0 exists");
        let param_name = captures.get(1).expect("group 1 exists").as_str();
        let Some(replacement) = replacements.get(param_name) else {
            bail!(
                "{}:{line_number}: destination references unknown parameter :{param_name}",
                redirects_path.display()
            );
        };
        translated.push_str(&destination[last_index..whole.start()]);
        translated.push_str(&format!("${replacement}"));
        last_index = whole.end();
    }
    translated.push_str(&destination[last_index..]);
    Ok(translated)
}

/// Minimal `shlex.split` (POSIX mode, comments off): whitespace-separated
/// tokens with single/double quotes and backslash escapes. Returns None on
/// unbalanced quotes or a trailing escape (Python raises ValueError).
fn shlex_split(line: &str) -> Option<Vec<String>> {
    let mut parts = Vec::new();
    let mut current = String::new();
    let mut in_word = false;
    let mut chars = line.chars();
    while let Some(ch) = chars.next() {
        match ch {
            c if c.is_whitespace() => {
                if in_word {
                    parts.push(std::mem::take(&mut current));
                    in_word = false;
                }
            }
            '\'' => {
                in_word = true;
                loop {
                    match chars.next() {
                        Some('\'') => break,
                        Some(inner) => current.push(inner),
                        None => return None,
                    }
                }
            }
            '"' => {
                in_word = true;
                loop {
                    match chars.next() {
                        Some('"') => break,
                        Some('\\') => match chars.next() {
                            // In double quotes, backslash only escapes the
                            // quote and itself (shlex posix behavior).
                            Some(escaped @ ('"' | '\\')) => current.push(escaped),
                            Some(other) => {
                                current.push('\\');
                                current.push(other);
                            }
                            None => return None,
                        },
                        Some(inner) => current.push(inner),
                        None => return None,
                    }
                }
            }
            '\\' => {
                in_word = true;
                let escaped = chars.next()?;
                current.push(escaped);
            }
            other => {
                in_word = true;
                current.push(other);
            }
        }
    }
    if in_word {
        parts.push(current);
    }
    Some(parts)
}

#[cfg(test)]
mod tests {
    //! Port of `tests/test_staticfile_provider.py`.

    use super::compute_redirects_config;

    #[test]
    fn test_staticfile_redirects_generate_sws_config_from_static_dir() {
        let tmp = tempfile::tempdir().unwrap();
        let static_dir = tmp.path().join("site");
        std::fs::create_dir_all(&static_dir).unwrap();
        std::fs::write(
            static_dir.join("_redirects"),
            "/docs/* /guides/:splat/ 301\n/blog/:slug /posts/:slug 302\n",
        )
        .unwrap();

        assert_eq!(
            compute_redirects_config(tmp.path(), Some("site"), true)
                .unwrap()
                .as_deref(),
            Some(
                "[[advanced.redirects]]\n\
                 source = \"/docs/{**}\"\n\
                 destination = \"/guides/$1/\"\n\
                 kind = 301\n\
                 \n\
                 [[advanced.redirects]]\n\
                 source = \"/blog/{*}\"\n\
                 destination = \"/posts/$1\"\n\
                 kind = 302\n"
            )
        );
    }

    #[test]
    fn test_staticfile_redirects_fall_back_to_project_root() {
        let tmp = tempfile::tempdir().unwrap();
        std::fs::create_dir_all(tmp.path().join("site")).unwrap();
        std::fs::write(
            tmp.path().join("_redirects"),
            "/docs/* /guides/:splat/ 301\n",
        )
        .unwrap();

        assert_eq!(
            compute_redirects_config(tmp.path(), Some("site"), true)
                .unwrap()
                .as_deref(),
            Some(
                "[[advanced.redirects]]\n\
                 source = \"/docs/{**}\"\n\
                 destination = \"/guides/$1/\"\n\
                 kind = 301\n"
            )
        );
    }

    #[test]
    fn test_staticfile_redirects_reject_unsupported_conditions() {
        let tmp = tempfile::tempdir().unwrap();
        std::fs::write(
            tmp.path().join("_redirects"),
            "/docs/* /guides/:splat/ 301 Country=us\n",
        )
        .unwrap();

        let error = compute_redirects_config(tmp.path(), None, true).unwrap_err();
        assert!(
            error
                .to_string()
                .contains("conditions and forced redirects are not supported"),
            "{error:#}"
        );
    }
}
